Shopify/shopify-cli
# Add static type checking with Sorbet

### WHY are these changes introduced?
Static type analysis can catch type-related issues that otherwise would lead to runtime errors. This allows shipping changes with more confidence and users get a more stable version of the CLI as a result. 

### WHAT is this pull request doing?
I'm introducing [Sorbet](https://sorbet.org/), and updating the CI pipeline to type-check in every commit. Note that most of the files include the `# typed: ignore` at the top to ignore them. The idea is that we incrementally adopt the tool as we are adding more code or modifying the existing one.

### How to test your changes?
1. Check out the branch.
2. Run `bundle install`.
3. Type-check with `bundle exec srb tc`

### Resources
- [Discourse thread asking for some benchmarking loading the runtime gem](https://discourse.shopify.io/t/performance-implications-when-enabling-the-sorbet-runtime-in-a-cli-project/24084)
- [Sorbet dashboard](https://sorbet-metrics.docs.shopify.io/)
- [Sorbet docs in the development handbook](https://development.shopify.io/engineering/keytech/reference/sorbet/other_projects)
